### PR TITLE
New UI

### PR DESCRIPTION
**Pull request by almirus**
image

jquery has been deleted

You need to implement 4 API 
/ble/bonding_db (GET)
/ble/bonding_db/whitelist (PUT, DELETE)
/ble/bonding_db/blacklist (PUT, DELETE)
/status (GET)

size was reduced about 30kb

### CODE REVIEW

File: www/js/ble2mqtt.js [hunk 1]
```diff
@@ -1,137 +1,298 @@
...
+function restart() {
+    fetch(RESTART, {
+        method: 'POST',
+    })
+        .then((res) => {
+            console.log("restart")
+            window.open("/busy.html", "_self")
+        })
+        .catch((err) => {
+            document.getElementById('message').innerText = "Can't restart";
```
Comment (shmuelzon) [inline www/js/ble2mqtt.js:38]: Shouldn't this be a call to `toaster()`?

File: www/js/ble2mqtt.js [hunk 2]
```diff
@@ -1,137 +1,298 @@
...
-    processData: false,
-  })
-  .done(onOtaUploadCompleted)
-  .fail(onOtaUploadCompleted);
+function restart() {
+    fetch(RESTART, {
+        method: 'POST',
+    })
+        .then((res) => {
+            console.log("restart")
```
Comment (shmuelzon) [inline www/js/ble2mqtt.js:34]: Please remove these debug logs

File: www/js/ble2mqtt.js [hunk 3]
```diff
@@ -1,137 +1,298 @@
...
+    progress(true);
+    fetch(OTA_UPLOAD + type, {
+        method: 'POST',
+        headers: {
+            'Content-Type': 'application/octet-stream'
+        },
+        body: file
+    })
+        .then((res) => {
+            document.getElementById('message').innerText = "Upload complete";
```
Comment (shmuelzon) [inline www/js/ble2mqtt.js:73]: Shouldn't this be a call to `toaster()`?

**Review by shmuelzon — CHANGES_REQUESTED** [review]
Hey,

Thanks for the pull request! I was hoping someone more up-to-date than I would be able improve the UI.

Regarding the new backend APIs, I understand the `/status` one just needs to return the current version but I'm not sure I understand the rest. The bonding DB is basically a list of paired device and the keys used to encrypt the traffic, there's no whitelist/blacklist there. Also, do you really need to the see the actual list? The API to clear it was added since sometimes the BLE peripheral is reset and its keys are removed forcing you to pair again, so the ESP needs to remove the device from its DB.

Also, can we do anything with the horizontal scroll bars?
image

Thanks!

File: www/js/ble2mqtt.js (hunk 1 above)
Comment (almirus) [reply www/js/ble2mqtt.js:38]: yep

File: www/js/ble2mqtt.js (hunk 3 above)
Comment (almirus) [reply www/js/ble2mqtt.js:73]: yep

File: www/index.html [hunk 4]
```diff
@@ -1,45 +1,97 @@
...
+                <div class="section"><label for="file-manager-file" class="button small primary"
+                                            title="Start upload to current dir">Upload</label>
+                    <input type="file" id="file-manager-file" style="display:none" onchange="uploadFile()"></div>
+                <div class="section">
+
+                    <table>
+                        <thead>
+                        <tr>
+                            <th>name</th>
+                            <th>size</th>
```
[the remark this answers is not in the document]
Comment (almirus) [reply www/index.html:52]: ok

**Comment by almirus** [comment]
> I understand the `/status`

now is only for get software version, hereafter you can return data on the amount of free memory for example.

> Also, do you really need to the see the actual list?

for this

>whitelist/blacklist - An array of MAC addresses of devices. If whitelist is used, only devices with a MAC address matching one of the entries will be connected while if blacklist is used, only devices that do not match any entry will be connected.

File: www/js/ble2mqtt.js (hunk 2 above)
Comment (almirus) [reply www/js/ble2mqtt.js]: done

File: www/index.html [hunk 5]
```diff
@@ -44,29 +44,17 @@ <h3 class="section">File manager</h3>
...
                     <input type="file" id="file-manager-file" style="display:none" onchange="uploadFile()"></div>
                 <div class="section">
-
                     <table>
                         <thead>
                         <tr>
-                            <th>name</th>
-                            <th>size</th>
+                            <th>NAME</th>
+                            <th>SIZE</th>
```
Comment (shmuelzon) [inline www/index.html:51]: Sorry for the misunderstanding, I meant only the first letter, i.e. `Name` and `Size`.

File: www/index.html [hunk 6]
```diff
@@ -80,13 +68,12 @@ <h3 class="section">File manager</h3>
         <table>
             <thead>
             <tr>
-                <th>Name</th>
+                <th>NAME</th>
                 <th>MAC</th>
-                <th>Actions</th>
+                <th>ACTIONS</th>
```
Comment (shmuelzon) [inline www/index.html]: These were fine the way they were

**Review by shmuelzon — CHANGES_REQUESTED** [review]

**Comment by shmuelzon** [comment]
> now is only for get software version, hereafter you can return data on the amount of free memory for example.

OK. Just wanted to make sure I wasn't missing anything else for the API I need to implement.

As for the bonding DB, it isn't the same as the configuration file's whitelist/blacklist. The bonding DB is internal to the BLE stack where it saves the encryption keys it created when it paired with BLE peripherals. The whitelist/blacklist is at the application level (in the configuration file) to tell the application which devices it should attempt to connect to and which to ignore. Even when we're configured to connect to a device, we might not even pair with it so it won't appear in the pairing DB.

So, the question is, what do you want to show in the UI. The list of devices we're connected to? The list in the pairing DB (not sure that's much of interest)? The list of detected devices? Even then, if you want to add devices to the white/black list you would need to edit the configuration file taking into account that only one list can actually be used. And if we're editing the configuration file, we can use the `/fs/` APIs instead of adding specific ones for manipulating the configuration file.

**Comment by almirus** [comment]
@shmuelzon my idea was to conveniently add to the black / white list
now I don't know if this functionality is needed at all...

**Comment by shmuelzon** [comment]
I think that, as a feature, it would be something nice to have. But it should be designed a bit differently. First of all, it's not related to the bonding DB.
I can add the following API:
`GET /ble/devices` - Will return a list of BLE devices in the following format:
```json
[
  {
    "mac_address": "aa:bb:cc:dd:ee:ff",
    "connected": true
  }
]
```
Those would be the bare minimum, I can add more fields if you want.
I would very much prefer that manipulation of the configuration file would be by the frontend as it would have more resources and, by using the `/fs/` routes, would not be limited it what it can do.
For adding/removing from the white/black list, the front end can get read the configuration file, make the necessary modifications and push an updated version. Please note that you cannot have both a whitelist or a blacklist so you would need to check what's configured (whitelist has precedence if both are configured) and the options should either be either add to the current list (if it's not currently there), or remove from the current list (if it is currently there).

I realize it's more work than what you've currently implemented, so we can do that as a separate commit and leave, for now, only the management section and the button for clearing the pairing DB.

As for the status API, do you want to add it to your commit, or should I commit it separately? 
```patch
diff --git a/main/httpd.c b/main/httpd.c
index c21d10c..4854f4f 100644
--- a/main/httpd.c
+++ b/main/httpd.c
@@ -45,6 +45,23 @@ static esp_err_t restart_handler(httpd_req_t *req)
     return ESP_OK;
 }
 
+static esp_err_t status_handler(httpd_req_t *req)
+{
+    esp_err_t ret;
+    char *response_str;
+    cJSON *response = cJSON_CreateObject();
+
+    cJSON_AddStringToObject(response, "version", BLE2MQTT_VER);
+
+    response_str = cJSON_PrintUnformatted(response);
+    httpd_resp_set_type(req, "application/json");
+    ret = httpd_resp_sendstr(req, response_str);
+
+    cJSON_free(response_str);
+    cJSON_Delete(response);
+    return ret;
+}
+
 static int register_management_routes(httpd_handle_t server)
 {
     httpd_uri_t uri_restart = {
@@ -53,8 +70,15 @@ static int register_management_routes(httpd_handle_t server)
         .handler  = restart_handler,
         .user_ctx = NULL,
     };
+    httpd_uri_t uri_status = {
+        .uri      = "/status",
+        .method   = HTTP_GET,
+        .handler  = status_handler,
+        .user_ctx = NULL,
+    };
 
     httpd_register_uri_handler(server, &uri_restart);
+    httpd_register_uri_handler(server, &uri_status);
 
     return 0;
 }
```

**Comment by almirus** [comment]
+ name column 
```
[
  {
    "mac_address": "aa:bb:cc:dd:ee:ff",
    "connected": true
    "name": "device_name"
  }
]

```

**Comment by almirus** [comment]
> do you want to add it to your commit

I don't know С well 😌

**Comment by shmuelzon** [comment]
No worries, I added the `/status` API to master.
Adding the name of the device is problematic for now since I don't read it at any point. We can add it to the API and it'll be an empty string for now and maybe I'll add it at some point.

**Comment by shmuelzon** [comment]
BTW - Did you see my comment from before regarding the horizontal scroll bars? I see them with Chrome.

**Comment by almirus** [comment]
> BTW - Did you see my comment from before regarding the horizontal scroll bars? I see them with Chrome.

already fixed here [www/css/mini-default.min.css](https://github.com/shmuelzon/esp32-ble2mqtt/pull/136/commits/eb05cd6f30e26291ba37c6ed809861df525148a0)

**Comment by shmuelzon** [comment]
Hey,

I pushed the new API for getting the BLE devices list, seems to work fine with your commit.
I just have 2 requests, if possible:
1. Please add an `<Unknown>` name or something when the name field is an empty string
2. Please add some visible marker on connected devices. I saw the tooltip, but it would be nice to quickly see what's connected

It seems a bit naked now :)
image

Thanks again!

**Comment by almirus** [comment]
hi! If you return "Unknown" for name property now, it is  fixed and also will be status shown via color                    
https://github.com/shmuelzon/esp32-ble2mqtt/blob/7af583b1e53c83d15c737a9dda6919e0d468a911/www/js/ble2mqtt.js#L159

**Comment by shmuelzon** [comment]
The `tertiary` class doesn't catch the `<td>` tag and doesn't affect the text color:
image

**Comment by almirus** [comment]
> The `tertiary` class doesn't catch the `<td>` tag and doesn't affect the text color
Oops

**Comment by shmuelzon** [comment]
If a device's name isn't available, would you prefer an empty string in the API or `null`?

**Comment by almirus** [comment]
> If a device's name isn't available, would you prefer an empty string in the API or `null`?

"not available" via API

**Comment by shmuelzon** [comment]
I don't agree with that. I think the API should properly reflect if a name is available or not and let the UI decide what to do with that. I can change that later.

**Comment by almirus** [comment]
do you want "<Unknown>"?

**Comment by shmuelzon** [comment]
Thanks again for the pull request! The new UI looks great!

**Pull request merged**